util/analyze-contention-log.sh

```diff
@@ -0,0 +1,148 @@
+#!/bin/sh
+
+###################################################################
+# Script to analyze logs produced by REPORT_RWLOCK_CONTENTION
+############################### ###################################
+
+#
+# Setup a temp dir to massage our contention log
+#
+TEMPDIR=$(mktemp -d /tmp/contention.XXXXXX)
+
+#
+# Specify our input log file
+#
+LOGFILE=$1
+
+#
+# Make sure we clean up on exit
+#
+trap "rm -rf $TEMPDIR" EXIT
+
+if [ ! -f $LOGFILE ]
+then
+    echo "No log file found"
+    exit 1
+fi
+LOGFILEBASE=$(basename $LOGFILE)
+
+echo "Splitting files"
+#
+# start by splitting the log into separate stack traces
+# the produced log file splits stack traces on empty lines
+# so we use the split command to break the big log up
+# into lots of little logs, so we can work with each
+# separately
+mkdir $TEMPDIR/individual_files
+cp $LOGFILE $TEMPDIR/individual_files/$LOGFILEBASE
+pushd $TEMPDIR/individual_files/ > /dev/null
+awk '
+    BEGIN {RS = ""; FS = "\n"}
+    {file_num++; print > ("stacktrace" file_num ".txt")}' ./$LOGFILEBASE
+popd > /dev/null
+rm $TEMPDIR/individual_files/$LOGFILEBASE
+
+#
+# Declare some associative arrays to track data for each file
+# each array is indexed by the sha1sum of the file
+#
+# We track the file name, the total latency accumulated by
+# that blocking event and the number of instances that we 
+# encountered it
+#
+declare -A filenames
+declare -A total_latency
+declare -A latency_counts
+
+echo "Gathering latencies"
+FILECOUNT=$(ls $TEMPDIR/individual_files/stacktrace*.* | wc -l)
+let currentidx=0
+
+#
+# Now look at every stack trace, get, and record its latency, and hash value
+#
+for i in $(ls $TEMPDIR/individual_files/stacktrace*.*)
+do
```

Comment:
I prefer `for x in y ; do` and similarly for `if` statements but that's just a matter of taste.